libraries/AP_NMEA_Output/AP_NMEA_Output.cpp

```diff
@@ -101,22 +101,25 @@ void AP_NMEA_Output::update()
 
     // format latitude
     char lat_string[13];
-    float deg = fabsf(loc.lat * 1.0e-7f);
+    double deg = fabsF(loc.lat * 1.0e-7f);
+    double min_dec = ((fabs(loc.lat) - (unsigned)deg * 1.0e7)) * 60 * 1.e-7f; 
     snprintf(lat_string,
              sizeof(lat_string),
              "%02u%08.5f,%c",
```

Comment:
I wonder if the problem could be the format string used.  The "%02u" looks ok but the "%08.5f" looks funny.  I'd guess that it should be "%8.7f" meaning 8 digits in total with 7 decimals of precisions.  E.g. , ".1234567"

Reply:
Good point, but I don't think it is the issue. According to my debugging, NMEA uses only those 5 decimal points. The complete conversion would be as follows:

- We have this lat long ( deg * 10e7 ) from ahrs.loc:

333807578 -1055973422

- The NMEA lat lon string corresponding to this would be:

lat: 3322.84547,N
lon: 10535.84053,W

It has "merged" degree and minutes, so lat would be 33 degree 22.84547 minutes North and so on.

From my debugging the issue came before this snprintf, as soon as we multiplied loc.lat or loc.lng by 1.0e-7f and load it into a double the rounding issue appeared, no matter if using fabsF, or fabs with ALLOW_DOUBLE_MATH_FUNCTIONS.

The last modification proposed ( I did a push force to maintain clean the tree ) seems to be working fine and seems to keep the same precision as the original ahrs.location info.

Thanks!

Reply:
I'm pretty sure you do need to use fabs and ALLOW_DOUBLE_MATH_FUNCTIONS
please test with sim_vehicle.py and --postype-single and --ekf-single, that will simulate one of our boards that uses single precision maths